java/arcs/core/storage/CapabilitiesResolverNew.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import arcs.core.common.ArcId
+import arcs.core.data.CapabilitiesNew
+import arcs.core.data.CollectionType
+import arcs.core.data.EntityType
+import arcs.core.data.ReferenceType
+import arcs.core.data.Schema
+import arcs.core.data.SingletonType
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.core.type.Type
+
+/**
+ * [CapabilitiesResolverNew] is a factory class that creates [StorageKey]s based on the registered
+ * providers and given [CapabilitiesNew].
+ */
+class CapabilitiesResolverNew(
+    val options: Options,
+    factoriesList: List<StorageKeyFactory> = listOf(),
+    val selector: FactorySelector = SimpleCapabilitiesSelector()
+) {
+    val factories: Map<String, StorageKeyFactory>
+
+    init {
+        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
+            "Storage keys protocol must be unique $factoriesList."
+        }
+        factories = factoriesList.map { it.protocol to it }.toMap().toMutableMap()
```

Comment:
`factoriesList.associateBy { it.protocol }.toMutableMap()`

Reply:
yes, thank you!